lib/using_restricted_array.rb

```diff
@@ -1,82 +1,140 @@
 def find_largest(array, length)
-  raise NotImplementedError
+  largest_number = array[0]
+  length.times do |i|
+    if array[i] > largest_number
+      largest_number = array[i]
+    end
+  end
+  return largest_number
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n)
+# Space complexity: O(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest_number = array[0]
+  length.times do |i|
+    if smallest_number > array[i]
+      smallest_number = array[i]
+    end
+  end
+  return smallest_number
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse(array, length)
-  raise NotImplementedError
+  n = 1
+  (length / 2).times do |i|
+    temp = 0
+    temp = array[i]
+    array[i] = array[length - n]
+    array[length - n] = temp
+    n += 1
+  end
+  return array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log n), logarithmic
+# Space complexity: O(1)
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  if length == 0
+    return false
+  end
+  low_point = 0
+  high_point = length - 1
+  while low_point < high_point
+    mid = (low_point + high_point) / 2
+    if array[mid.floor] == value_to_find
+      return true
+    elsif array[mid.floor] > value_to_find
+      high_point = mid - 1
+    elsif low_point = mid + 1
+    end
+  end
+  if low_point < high_point
+    return false
+  end
+  if array[low_point] == value_to_find
+    return true
+  else
+    return false
+  end
 end
 
 # Helper method provided to sort the array in ascending order
 # Implements selection sort
-# Time complexity = O(n^2), where n is the number of elements in the array.
+# Time complexity = O(n^2), where n is the value_to_find ofarray in the array.
 # To find the correct value for index 0, a total of (n-1) comparisons are needed.
 # To find the correct value for index 1, a total of (n-2) comparisons are needed.
 # To find the correct value for index 2, a total of (n-3) comparisons are needed.
 # and so on ...
 # To find the correct value for index (n-2), a total of 1 comparisons is needed.
 # To find the correct value for the last index, a total of 0 comparisons are needed.
-# Total number of comparisons = (n-1) + (n-2) + ... 3 + 2 + 1
+# Total value_to_find of comparisons = (n-1) + (n-2) + ... 3 + 2 + 1
 #                             = (n * (n-1))/2
 # This is O(n^2) in Big O terms.
 # Space complexity = constant or O(1) since the additional storage needed,
 # does not depend on input array size.
 def sort(array, length)
-  length.times do |index| # outer loop - n elements
+  length.times do |index| # outer loop - narray
```

Comment:
Looks like you did find/replace of `"number"` -> `"value_to_find"` and `" elements"` -> `"array"`? 😁